### PR TITLE
Translate the repetition pattern to expand and reshape

### PR DESCRIPTION
**Pull request by naoyam**
In RoPE, this repeat pattern shows up commonly:

https://github.com/huggingface/transformers/blob/main/src/transformers/models/llama/modeling_llama.py#L136
```
torch.cat((freqs, freqs), dim=-1)
```

This pattern shows up as `PadOp` followed by `CatOp` in nvFuser.

This preseg pass translates this pattern to expand and reshape ops. For example, given a pattern like:

```
t0 = [i0];
t1 = cat({t0, t0}, -1);
```

It will be translated to:

```
t0 = [i0]
t2 = broadcast(t0, {true, false});
t3 = expand(t2, {2, i0});
t4 = reshape(t3, {2 * i0});
```

And all uses of `t1` will be replaced by `t4`.

While the pattern can be handled by the resize scheduler, it's currently limited to segments with pointwise ops only, and its scheduling heuristics are not tuned yet. Specifically, I experimentally observed a significant perf gain with the Mistral backward function.

### CODE REVIEW

**Comment by naoyam** [comment]
!test --diff

**Comment by naoyam** [comment]
!test --diff

**Comment by naoyam** [comment]
!test --diff

File: csrc/preseg_passes/translate_repeat_to_expand.cpp [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+
+ private:
+  // Traverse through the fusion and gather all patterns of a pad
+  // followed by a concat. If a single concat op has multiple pad
+  // inputs that resize the same iter domain of the same input tensor,
+  // that must correspond to a repetition.
+  void inspect() {
+    const auto exprs = fusion_->exprs();
+
+    for (auto pad : ir_utils::filterByType<PadOp>(exprs)) {
```
Comment (jjsjann123) [inline csrc/preseg_passes/translate_repeat_to_expand.cpp:55]: QQ: what's the reason that we choose to filter on `PadOp` here, instead of looking for `CatOp` and iterate through all its input?

It's just a nitpicking question, I think when we iterate through each `PadOp`, we could repetitively remove entries and adding things back into `repeat_info_map_`.

File: csrc/preseg_passes/translate_repeat_to_expand.cpp [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+          inp_domain.begin(),
+          std::find(inp_domain.begin(), inp_domain.end(), info.repeated_id));
+      bcast_flags.at(repeated_id_offset) = true;
+      auto broadcast_tv = broadcast(info.input_tv, bcast_flags);
+      NVF_ERROR(broadcast_tv->nDims() == inp_domain.size() + 1);
+
+      // Step 2
+      std::vector<Val*> expanded_sizes(
+          bcast_flags.size(), IrBuilder::create<Val>(-1L));
+      expanded_sizes.at(repeated_id_offset) = IrBuilder::create<Val>(2L);
```
Comment (jjsjann123) [inline csrc/preseg_passes/translate_repeat_to_expand.cpp:168]: This shouldn't be a hard-coded 2 here. I think our logic can handle something like `cat({t0, t0, t0}, -1)` as well and we don't have a check to exclude that neither.
We should use the # of inputs here instead.

Comment (jjsjann123) [reply csrc/preseg_passes/translate_repeat_to_expand.cpp:168]: ^^^ Let's also add a test case with that.

File: csrc/preseg_passes/translate_repeat_to_expand.cpp (hunk 1 above)
Comment (naoyam) [reply csrc/preseg_passes/translate_repeat_to_expand.cpp:55]: I think I was initially also thinking about supporting concatenation without `CatOp`. I think I have seen a sequence of ops like a `PadOp` followed by an addition, maybe in some backward kernels. 

I dropped that since I don't find that pattern for repetitions. 
https://github.com/NVIDIA/Fuser/pull/3645/commits/4c1abda909fbf2d85786b13f3d1566a3490db567

File: csrc/preseg_passes/translate_repeat_to_expand.cpp (hunk 2 above)
Comment (naoyam) [reply csrc/preseg_passes/translate_repeat_to_expand.cpp:168]: Ah, that's a very good point. Yes, I took a shortcut and forgot to update it.

File: tests/cpp/test_preseg_passes.cpp [hunk 3]
```diff
@@ -770,4 +771,175 @@ TEST_F(PresegTest, DisjointSetsOfExtentsConcreteSymbolic) {
...
+
+  auto tv0 = makeContigConcreteTensor({4, 8});
+  fusion.addInput(tv0);
+
+  // Consecutive repetitions with different IDs
+  auto tv1 = cat({tv0, tv0}, 1);
+  auto tv2 = cat({tv0, tv0}, 1);
+
+  fusion.addOutput(tv1);
+  fusion.addOutput(tv2);
```
[the remark this answers is not in the document]
Comment (naoyam) [reply tests/cpp/test_preseg_passes.cpp:916]: First of all, the comment at line 911 is wrong. Both cat ops use the same ID.

Yes, `tv1` and `tv2` are disconnected. No property that we use to build IdModel would detect the two IDs of the tensors have the same extent. It doesn't mean we can't extend IdModel to detect that, it's just not how it's implemented at this moment. Unless there's any motivating case, I don't think we would need to worry too much about it.

**Comment by naoyam** [comment]
!test

**Review by jjsjann123 — APPROVED** [review]
LGTM

**Comment by naoyam** [comment]
!build
